src/DataProtection/DataProtection/src/KeyManagement/Key.cs

```diff
@@ -16,7 +17,13 @@ namespace Microsoft.AspNetCore.DataProtection.KeyManagement;
 /// </summary>
 internal sealed class Key : IKey
 {
-    private readonly Lazy<IAuthenticatedEncryptorDescriptor> _lazyDescriptor;
+    private IAuthenticatedEncryptorDescriptor? _descriptor;
+
+    // If descriptor is available at construction time, these will remain null forever
+    private readonly object? _descriptorLock; // Protects _descriptor and _descriptorException
```

Comment:
```suggestion
	[MemberNotNullWhen(false, nameof(_descriptor)]
    private object? DescriptorLock { get; }  // Protects _descriptor and _descriptorException
```

Nit: make this a property and codify the mutually exclusive constraints with `MemberNotNullWhen`? Ditto for `_descriptorException`.

Reply:
I was sure there was such an attribute...

Reply:
CS0592	Attribute 'MemberNotNullWhen' is not valid on this declaration type. It is only valid on 'method, property, indexer' declarations.

Reply:
Oh, I missed "make this a property".  Really?  Just for the attribute?

Reply:
Nope, still not getting it.  Don't you have to apply that attribute to something bool-returning?

Reply:
> Oh, I missed "make this a property". Really? Just for the attribute?

Yes. :/ Main reason I recommended it is because I think it's a little cleaner than Debug.Assert's sprinkled throughout the code.

> Nope, still not getting it. Don't you have to apply that attribute to something bool-returning?

Nope, I believe the first argument is the list of the member referenced in the member being `== null`.

Reply:
Per offline discussion, it looks like using the attribute isn't useful anyways since the constraint that "_descriptorLock and _descriptorExceptionwill always be null if_descriptor is not null" is not true.

Reply:
Sticking with comments and asserts for now.